File: docker/host/programs/mult_send.c

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <netinet/in.h>
#include <netdb.h>
#include <ctype.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>

const int TTL = 20;
const int BUFFSIZE = 512;
int udp_sock = -1;
int run = 1;


/*
 * SIGTERM Handler
 */
void close_handler(int sig) {
    run = 0;
}


/*
 * Main function
 */
int main(int argc, char **argv) {
    int c;
    int ip_addr = htonl(0xef010101); // 239.1.1.1
    int port = 52220;

    char buffer[BUFFSIZE];
    struct sockaddr_in remote_addr;
    int addr_length = sizeof(remote_addr);

    /*
     * params
     *   -s : use SSM address instead of SM address
     */
    while ((c=getopt(argc,argv,"s")) != -1) {
        switch(c) {
            case 's':
                ip_addr = htonl(0xe8010101); // 232.1.1.1
	        break;
        }
    }

    // prepare remote address struct
    remote_addr.sin_addr.s_addr = ip_addr;
    remote_addr.sin_family=AF_INET;
    remote_addr.sin_port=htons(port);

    // create udp socket
    if ((udp_sock=socket(AF_INET,SOCK_DGRAM,0))==-1) {
        perror("Unable to create udp socket");
        exit(1);
    }

    // catch signals
    signal (SIGINT, close_handler);
    signal (SIGTERM, close_handler);
    signal (SIGKILL, close_handler);

    // send udp multicast paquet
    setsockopt(udp_sock, IPPROTO_IP, IP_MULTICAST_TTL, &TTL, sizeof(TTL));
    while(run) {
        sendto(udp_sock,buffer,BUFFSIZE, 0,
                (const struct sockaddr*) &remote_addr, addr_length);
        sleep(1);
    }

    // close the socket
    if (close(udp_sock) == -1) {
        perror("Unable to close udp socket");
        exit(1);
    }

    return(0);
}

File: docker/host/programs/mult-send.py
#!/usr/bin/env python3

import socket
import argparse
import logging
import time
import signal

PORT = 5222
TTL = 64
RUNNING = True

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Multicast sending program')
    parser.add_argument(
        "-g", "--group", type=str, dest="group",
        metavar="IP_ADDR", default="239.1.1.1",
        help="Multicast IP address destination")
    parser.add_argument(
        "-i", "--interval", type=float, dest="interval",
        metavar="INTERVAL", default=1.0,
        help="Interval between 2 packets")
    args = parser.parse_args()
    
    log_format = '%(levelname)s: %(message)s'
    logging.basicConfig(format=log_format, level=logging.INFO)

    sock = socket.socket(socket.AF_INET,
                        socket.SOCK_DGRAM,
                        socket.IPPROTO_UDP)
    sock.setsockopt(socket.IPPROTO_IP,
                    socket.IP_MULTICAST_TTL,
                    TTL)
    
    def handler(_signum, _frame):
        global RUNNING

        logging.info("Stop sending...")
        RUNNING = False

    signal.signal(signal.SIGINT, handler)
    signal.signal(signal.SIGTERM, handler)

    logging.info(f"Send multicast packets to group {args.group}")
    counter = 1
    while RUNNING:
        logging.info(f"Send packet {counter}")
        sock.sendto(f"Multicat Packet {counter}".encode('utf-8'), (args.group, PORT))
        counter += 1
        time.sleep(args.interval)
    
    sock.close()
